src/anemoi/inference/runner.py

```diff
@@ -391,186 +235,121 @@ def run(
-
-        LOGGER.info("Using autocast %s", autocast)
-
-        # Write dynamic fields
-        def get_most_recent_datetime(input_fields):
-            datetimes = [f.datetime()["valid_time"] for f in input_fields]
-            latest = datetimes[-1]
-            for d in datetimes:
-                assert d <= latest, (datetimes, d, latest)
-            return latest
-
-        most_recent_datetime = get_most_recent_datetime(input_fields)
-        reference_fields = [f for f in input_fields if f.datetime()["valid_time"] == most_recent_datetime]
-
-        if "lsm" in self.checkpoint.variable_to_index:
-            prognostic_template = reference_fields[self.checkpoint.variable_to_index["lsm"]]
-        else:
-            first = list(self.checkpoint.variable_to_index.keys())
-            LOGGER.warning("No LSM found to use as a GRIB template, using %s", first[0])
-            prognostic_template = reference_fields[0]
-
-        accumulated_output = np.zeros(
-            shape=(len(diagnostic_output_mask), number_of_grid_points),
-            dtype=np.float32,
-        )
+        ).to(self.device)
 
-        if self.checkpoint.diagnostic_params:
-            output_callback(
-                input_fields,
-                self.checkpoint.diagnostic_params,
-                prognostic_template,
-                accumulated_output[0].shape,
-            )
-        else:
-            output_callback(input_fields)
+        LOG.info("Using autocast %s", self.autocast)
 
-        prognostic_params = self.checkpoint.prognostic_params
-        accumulations_params = self.checkpoint.accumulations_params
+        lead_time = frequency_to_timedelta(lead_time)
+        steps = lead_time // self.checkpoint.frequency
 
-        # with self.stepper(self.hour_steps) as stepper:
+        LOG.info("Lead time: %s, frequency: %s Forecasting %s steps", lead_time, self.checkpoint.frequency, steps)
 
-        for i in progress_callback(range(lead_time // self.hour_steps)):
-            step = (i + 1) * self.hour_steps
+        result = input_state.copy()
+        result["fields"] = dict()
+
+        start = input_state["date"]
+
+        for i in range(steps):
+            step = (i + 1) * self.checkpoint.frequency
+            date = start + step
+            LOG.info("Forecasting step %s (%s)", step, date)
+
+            result["date"] = date
 
             # Predict next state of atmosphere
-            with torch.autocast(device_type=device, dtype=autocast):
-                y_pred = model.predict_step(input_tensor_torch)
+            with torch.autocast(device_type=self.device, dtype=self.autocast):
+                y_pred = self.model.predict_step(input_tensor_torch)
 
-            # Detach tensor and squeeze
-            output = np.squeeze(y_pred.cpu().numpy())
+            # Detach tensor and squeeze (should we detach here?)
+            output = np.squeeze(y_pred.cpu().numpy())  # shape: (values, variables)
```

Comment:
I think detaching should be an optional behaviour

Reply:
Let's merge and refactor in another PR